### PR TITLE
[Refactor] Fetch COW rewards from order_rewards table

### PR DESCRIPTION
**Pull request by bh2smith**
This PR implements the recently introduced "Risk Adjusted Solver Rewards" in [CIP-14](https://snapshot.org/#/cow.eth/proposal/0x3c84fc8e3cfe9cc6df76198d5031fae6580d8f9531f5b92ca3adedbc976cb1e5)

To summarize:

- we now fetch the cow rewards from the `order_rewards` table in the orderbook database
- validation is performed on the solver/recipient set and the trade counts retrieved from the orderbook and EVM event data (via Dune Analytics).

⚠️ This is **NOT TO BE MERGED UNTIL AFTER** the solver payout transaction execution on October 18, 2022! This is also pending the results of the DAO vote linked above.

## Test Plan

End to end test for the get_cow_rewards method has been adapted to check that our validation does not fail. Unfortunately this test can't legitimately pass now because it requires data to be contained in both production and staging databases. Once this goes live we can change the accounting period dates to match and the results test should yield passing assertions.

Specifically, testing this PR should amount to the passing of the following e2e test file:

```
python -m pytest tests/e2e/test_get_transfers.py
```

### CODE REVIEW

File: queries/cow_rewards.sql [hunk 1]
```diff
@@ -10,59 +10,17 @@ with trade_hashes as (SELECT solver,
...
-                                       num_trades,
-                                       num_trades - num_orders as missing_orders
-                                from db_order_counts tc
-                                         inner join lateral (
-                                    select count(order_uid) as num_trades,
-                                           tx_hash          as hash
-                                    from trade_hashes
-                                    group by tx_hash
-                                    ) as th on tc.tx_hash = th.hash),
+                      where block_number between 15719995 and 15727058)
```
Comment (josojo) [inline queries/cow_rewards.sql:15]: Why is it still hardcoded?

File: queries/cow_rewards.sql [hunk 2]
```diff
@@ -10,59 +10,17 @@ with trade_hashes as (SELECT solver,
...
---                           group by solver)
---
--- select '0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB'                        as token_address,
---        receiver,
---        ((50 * num_batches + 35 * num_trades) * pow(10, 18))::numeric::text as amount
--- from solver_solutions
--- order by receiver;
\ No newline at end of file
+select concat('0x', encode(solver, 'hex'))          as receiver,
+       -- This column will be used to compare with Dune:
```
Comment (josojo) [inline queries/cow_rewards.sql]: I really like that we have some sanity checks build int

**Review by josojo — APPROVED** [review]
Nice... looks ready!

But I wanna test it a little bit more with real data next week, before we merge it.

File: queries/cow_rewards.sql (hunk 1 above)
Comment (bh2smith) [reply queries/cow_rewards.sql:15]: Accident.

**Comment by bh2smith** [comment]
> But I wanna test it a little bit more with real data next week, before we merge it.

Yes of course. We absolutely cannot merge this until after the next payout. I guess its probably time to run tagged versions instead of main.

File: queries/cow_rewards.sql (hunk 1 above)
Comment (bh2smith) [reply queries/cow_rewards.sql:15]:
```suggestion
                      where block_number between {{start_block}} and {{end_block}})
```

**Comment by bh2smith** [comment]
Also, @josojo - we should see if Dune wants to index a test net like goerli so we could run tests on an actual test net.

**Comment by josojo** [comment]
>
Also, @josojo - we should see if Dune wants to index a test net like goerli so we could run tests on an actual test net.

I don't think it's super important. For which use cases, do you see a need for it?

**Comment by bh2smith** [comment]
Well generally we would run e2e tests on a test network (like any other project). However for all our dune projects we have no choice but to run mainnet queries with and mainnet other things. Anyway, its not all that important but would be nice not to cross wires (Right now we test the safe multisend on goerli but based on mainnet orderbook data).

File: queries/cow_rewards.sql [hunk 3]
```diff
@@ -10,59 +10,17 @@ with trade_hashes as (SELECT solver,
...
+       -- cross referencing that all trades are accounted for in the orderbook
+       count(*) as num_trades,
+       (sum(reward) * pow(10, 18))::numeric::text   as amount,
+       '0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB' as token_address
+from trade_hashes
+         -- Inner join because both prod and staging DB index trades and settlements,
+-- but the rewards for each environment are contained only in respective databases.
+         inner join order_rewards
+                    on trade_hashes.order_uid = order_rewards.order_uid
```
Comment (josojo) [inline queries/cow_rewards.sql:26]: Don't we also need to join on the auction_id? The order_uid is not unique, if we don't have constant orders rewards, right?

Comment (bh2smith) [reply queries/cow_rewards.sql:26]: Good thinking. I'll try it out and see.

Comment (bh2smith) [reply queries/cow_rewards.sql:26]: Is this what you has in mind?

https://github.com/cowprotocol/solver-rewards/pull/96/commits/981e6bdbca963921071300c603b35718cedb0189

File: queries/dune_trade_counts.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+group by solver
+
+-- V2 Query: https://dune.com/queries/1393633?d=1
+-- select solver,
+--        count(*) as num_trades
+-- from gnosis_protocol_v2_ethereum.GPv2Settlement_evt_Trade t
+--          join gnosis_protocol_v2_ethereum.GPv2Settlement_evt_Settlement s
+--               on t.evt_tx_hash = s.evt_tx_hash
+-- where t.evt_block_number between {{start_block}} and {{end_block}}
+-- group by solver
```
Comment (gentrexha) [inline queries/dune_trade_counts.sql:17]: What do you think about adding a TODO here explaining why we still left this as commented in the code?

File: src/fetch/transfer_file.py [hunk 5]
```diff
@@ -370,15 +371,45 @@ def get_cow_rewards(start_block: str, end_block: str) -> list[Transfer]:
...
+            name="Trade Counts",
+            parameters=[
+                QueryParameter.text_type("start_block", start_block),
+                QueryParameter.text_type("end_block", end_block),
+            ],
+        )
+    )
+    # Validation of results - using characteristics of results from two sources.
+    # Solvers do not appear in both environments!
+    assert set(prod_df.receiver).isdisjoint(set(barn_df.receiver)), "receiver overlap!"
```
Comment (gentrexha) [inline src/fetch/transfer_file.py:400]: This is really cool thing!

**Review by gentrexha — COMMENTED** [review]
Great work Ben! 

Left one minor comment, other than that very good job on this one!

**Comment by gentrexha** [comment]
I just realized this was still ongoing from yesterday from my side and it has been already solved. All is good.

File: queries/dune_trade_counts.sql (hunk 4 above)
Comment (bh2smith) [reply queries/dune_trade_counts.sql:17]: This is a good idea although from this project's perspective its not really a todo. Its a TODO from the Dune Migration perspective and this commented code is more of a gift to the person who actually does it... I could put a TODO here but its already merged and I don't see this happening any time soon.

**Comment by bh2smith** [comment]
> I just realized this was still ongoing from yesterday from my side and it has been already solved. All is good.

Not sure I understand this comment. What was ongoing?

File: src/fetch/transfer_file.py (hunk 5 above)
Comment (bh2smith) [reply src/fetch/transfer_file.py:400]: Glad you like it!

**Comment by gentrexha** [comment]
> > I just realized this was still ongoing from yesterday from my side and it has been already solved. All is good.
> 
> Not sure I understand this comment. What was ongoing?

I left this PR open over night and hand't seen that it was already merged.

File: queries/cow_rewards.sql (hunk 1 above)
Comment (bh2smith) [reply queries/cow_rewards.sql:15]: I see... I guess the suggestion was never committed...

Comment (bh2smith) [reply queries/cow_rewards.sql:15]: Fixed here: https://github.com/cowprotocol/solver-rewards/pull/104

**Pull request merged**